source/fundamentals/class-mapping.txt

```diff
@@ -0,0 +1,329 @@
+.. _csharp-class-mapping:
+
+=============
+Class Mapping
+=============
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to customize the way the {+driver-long+}
+maps BSON documents to and from {+language+} classes. You should read this page
+to learn more about the default class mapping behavior, or if you need to
+customize the way the driver serializes or deserializes your data.
+
+Automatic Class Mapping
+-----------------------
+
+When you use a class, rather than a ``BsonDocument``, to represent data in a 
+MongoDB collection, the {+driver-short+} automatically creates a **class map**
+that it uses to serialize or deserialize your data. It does this mapping by matching 
+the name of the field in the document to the name of the property in the class.
+
+.. important::
+
+   The type of the property in your class should match the type of the field in
+   the document. The {+driver-short+} instantiates a serializer based on the
+   type of the property in your class. If the types don't match when the driver
+   attempts to deserialize the data, the serializer throws an exception.
+
+Manually Creating A Class Map
+-----------------------------
+
+You can bypass the driver's automatic class mapping functionality, and manually
+define the class map by using the ``RegisterClassMap()`` method.
+
+The following example defines a ``Person`` class:
+
+.. code-block:: csharp
+
+   public class Person
+   {
+       public string Name { get; set; }
+       public int Age { get; set; }
+       public List<string> Hobbies {get; set;}
+   }
+
+The following code demonstrates how to register the class map for the ``Person``
+class: 
+
+.. code-block:: csharp
+   
+   BsonClassMap.RegisterClassMap<Person>(classmap => 
+   {
+       classmap.MapMember(p => p.Name);
+       classmap.MapMember(p => p.Age);
+       classmap.MapMember(p => p.Hobbies);
+   });
+
+.. important::
+
+   You must register a class map *before* it's needed in your code. We recommend
+   registering class maps prior to initializing a connection with MongoDB.
+
+You can also manually map a subset of class properties, while still
+allowing the driver to automatically map the remaining properties. To do this, 
+register a class map and call the ``AutoMap()`` method before manually 
+specifying your properties.
+
+In the following code example, the ``AutoMap()`` method maps all properties
+of the ``Person`` class except ``Hobbies``, which is mapped manually:
```

Comment:
This is not quite right. Automap still maps all properties (including Hobbies).

What you can do is alter the mapping after Automap runs.

This example is a no-op because `classMap.MapMember(p => p.Hobbies)` does nothing because `Hobbies` is already mapped.

Reply:
For this example to be meaningful the line after `AutoMap` should do something do change the mapping that `AutoMap` produced.

For example:

    classMap.MapMember(p => p.Hobbies).SetElementName("hobbies");

which of course is also a no-op if you assume a non-default set of conventions with camel case registered (which still concerns me), so maybe:

    classMap.MapMember(p => p.Hobbies).SetElementName("favorite_hobbies");

Reply:
Thanks, that does seem like a better example.

I've updated the wording for the last sentence before the example Let me know if it is more correct that way